collector/lib/KernelDriver.cpp

```diff
@@ -0,0 +1,138 @@
+#define delete_module(name, flags) syscall(__NR_delete_module, name, flags)
+
+namespace collector {
+
+namespace {
+
+std::string constructArgsStr(std::unordered_map<std::string, std::string>& args) {
+  bool first = true;
+  std::stringstream stream;
+
+  for (auto& entry : args) {
+    if (first) {
+      first = false;
+    } else {
+      stream << " ";
+    }
+    stream << entry.first << "=" << entry.second;
+  }
+
+  return stream.str();
+}
+
+int doInsertModule(std::vector<char>& image, std::unordered_map<std::string, std::string>& args) {
+  std::string args_str = constructArgsStr(args);
+  CLOG(DEBUG) << "Kernel module arguments: " << args_str;
+
+  int res = init_module(&image.front(), image.size(), args_str.c_str());
+  if (res != 0) {
+    return res;
+  }
+
+  std::string param_dir = GetHostPath(
+      std::string("/sys/module/") + SysdigService::kModuleName + "/parameters/");
+
+  struct stat st;
+  if (stat(param_dir.c_str(), &st) != 0) {
+    // This is not optimal, but don't fail hard on systems where
+    // for whatever reason the above directory does not exist.
+    CLOG(WARNING) << "Could not stat " << param_dir << ": " << StrError()
+                  << ". No parameter verification can be performed.";
+    return 0;
+  }
+
+  for (const auto& entry : args) {
+    std::string param_file = param_dir + entry.first;
+
+    if (stat(param_file.c_str(), &st) != 0) {
+      CLOG(ERROR) << "Could not stat " << param_file << ": " << StrError()
+                  << ". Parameter " << entry.first
+                  << " is unsupported, suspecting module version mismatch.";
+      errno = EINVAL;
+      return -1;
+    }
+  }
+  return 0;
+}
+
+}  // namespace
+
+bool KernelDriverModule::insert(const std::vector<std::string>& syscalls, std::string path) {
+  std::unordered_map<std::string, std::string> module_args;
+  std::stringstream syscall_ids;
+
+  // Iterate over the syscalls that we want and pull each of their ids.
+  // These are stashed into a string that will get passed to init_module
+  // to insert the kernel module
+  const EventNames& event_names = EventNames::GetInstance();
+  for (const auto& syscall : syscalls) {
+    for (ppm_event_type id : event_names.GetEventIDs(syscall)) {
+      syscall_ids << id << ",";
+    }
+  }
+  syscall_ids << "-1";
+
+  module_args["s_syscallIds"] = syscall_ids.str();
+  module_args["exclude_initns"] = "1";
+  module_args["exclude_selfns"] = "1";
+  module_args["verbose"] = "0";
+
+  std::ifstream module(path, std::ios_base::binary);
+
+  if (!module.is_open()) {
+    CLOG(ERROR) << "Cannot open kernel module: " << path << ". Aborting...";
+    return false;
+  }
+
+  module.seekg(0, std::ios_base::end);
+  size_t image_size = module.tellg();
+  module.seekg(0, std::ios_base::beg);
+
+  std::vector<char> image;
+  image.reserve(image_size);
+
+  std::copy(std::istreambuf_iterator<char>(module),
+            std::istreambuf_iterator<char>(),
+            std::back_inserter(image));
```

Comment:
@kcrane - I have adjusted the kernel module reading code to be a little more idiomatic C++, but I believe it behaves the same way as before. Could you have a look in case I've missed something that might affect s390x?

Reply:
all integration tests pass on s390x with your branch